src/box/memtx_tree.cc

```diff
@@ -423,9 +483,13 @@ name(struct iterator *iterator, struct tuple **ret)				\
 	struct space *space = space_by_id(iterator->space_id);			\
 	bool is_rw = txn != NULL;						\
 	do {									\
-		int rc = name##_base<USE_HINT>(iterator, ret);			\
-		if (rc != 0 || *ret == NULL)					\
-			return rc;						\
+		int rc = name##_base<USE_HINT>(iterator, ret);       		\
+		if (rc != 0 || *ret == NULL) {         				\
+			if (rc == 0)						\
+				memtx_read_iterator_track_read(txn, it,		\
```

Comment:
Here and anywhere this function can fail and return -1. I think that this case should be handled correctly.